join on OnClauseWrapper

<!--
If you want to report a bug, we added some points below you can fill out. If you want to request a feature, feel free to remove all the irrelevant text. In addition to this [issue tracker](https://github.com/diesel-rs/diesel/issues), you can also talk to Diesel maintainers and users on [Gitter](https://gitter.im/diesel-rs/diesel).
-->

## Setup

### Versions

- **Rust:**
rustc 1.37.0 (eae3437df 2019-08-13)
- **Diesel:**
1.4.2
- **Database:**
mariadb 15.1
- **Operating System**
Debian buster 

### Feature Flags

- **diesel:**
features = ["mysql", "chrono"] 

## Problem Description
It seems we can't do a join on an explicit OnClauseWrapper.

@weiznich advise me to open an issue for this problem. 

Feel free to ask if you need more information.

### What are you trying to accomplish?
An equivalent of this SQL query:
```
select * 
from table1
inner join table2 on table1.id = table2.t1_id and table2.type = "my_type"
left outer join table3 on table2.id = table3.t2_id
```

For that I'm doing:
```
table1.inner_join(table2.ON(...).left_outer_join(table3))
```
<!-- ### What is the expected output? -->

### What is the actual output?
I get this error message:
```
error[E0599]: no method named `left_outer_join` found for type `diesel::query_source::joins::OnClauseWrapper<schema::glpi_items_operatingsystems::table, diesel::expression::operators::And<diesel::expression::operators::Eq<schema::glpi_computers::columns::id, schema::glpi_items_operatingsystems::columns::items_id>, diesel::expression::operators::Eq<schema::glpi_items_operatingsystems::columns::itemtype, diesel::expression::bound::Bound<diesel::sql_types::Nullable<diesel::sql_types::Text>, &str>>>>` in the current scope
  --> src/queries.rs:91:14
   |
91 |             .left_outer_join(glpi_operatingsystems::table.on(glpi_items_operatingsystems::operatingsystems_id.eq(glpi_operatingsystems::id)))
   |              ^^^^^^^^^^^^^^^
   |
   = note: the method `left_outer_join` exists but the following trait bounds were not satisfied:
           `&diesel::query_source::joins::OnClauseWrapper<schema::glpi_items_operatingsystems::table, diesel::expression::operators::And<diesel::expression::operators::Eq<schema::glpi_computers::columns::id, schema::glpi_items_operatingsystems::columns::items_id>, diesel::expression::operators::Eq<schema::glpi_items_operatingsystems::columns::itemtype, diesel::expression::bound::Bound<diesel::sql_types::Nullable<diesel::sql_types::Text>, &str>>>> : diesel::QueryDsl`
           `&mut diesel::query_source::joins::OnClauseWrapper<schema::glpi_items_operatingsystems::table, diesel::expression::operators::And<diesel::expression::operators::Eq<schema::glpi_computers::columns::id, schema::glpi_items_operatingsystems::columns::items_id>, diesel::expression::operators::Eq<schema::glpi_items_operatingsystems::columns::itemtype, diesel::expression::bound::Bound<diesel::sql_types::Nullable<diesel::sql_types::Text>, &str>>>> : diesel::QueryDsl`
           `diesel::query_source::joins::OnClauseWrapper<schema::glpi_items_operatingsystems::table, diesel::expression::operators::And<diesel::expression::operators::Eq<schema::glpi_computers::columns::id, schema::glpi_items_operatingsystems::columns::items_id>,
diesel::expression::operators::Eq<schema::glpi_items_operatingsystems::columns::itemtype, diesel::expression::bound::Bound<diesel::sql_types::Nullable<diesel::sql_types::Text>, &str>>>> : diesel::QueryDsl`

```

<!-- ### Are you seeing any additional errors? -->

### Steps to reproduce
I did a small poc for reproducing the error:

main.rs
```
#[macro_use]
extern crate diesel;

use diesel::prelude::*;


table! {
    table1 (id) {
        id -> Integer,
    }
}
table! {
    table2 (id) {
        id -> Integer,
        t1_id -> Integer,
        t2_type -> Nullable<Varchar>,
    }
}
table! {
    table3 (id) {
        id -> Integer,
        t2_id -> Integer,
    }
}
allow_tables_to_appear_in_same_query!(
    table1,
    table2,
    table3
);

fn main() {
    let results = table1::table
        .inner_join(table2::table.on(table1::id.eq(table2::t1_id).and(table2::t2_type.eq("my_type")))
                    .left_outer_join(table3::table)
        )
        .select((table1::id, table2::id));

        println!("Hello, world!");
}

```
Cargo.toml
```
[package]
name = "join-poc"
version = "0.1.0"
authors = ["Yann <yann@fery.me>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
diesel = { version = "1.4.2", features = ["mysql", "chrono"] }
```
## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) for similar issues.
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)

<!--
Thank you for your submission!  You're helping make Diesel more robust 🎉

We'll try to respond as quickly as possible.
-->

--- a/diesel/src/query_dsl/join_dsl.rs
+++ b/diesel/src/query_dsl/join_dsl.rs
@@ -5,7 +5,7 @@ use crate::query_source::{JoinTo, QuerySource, Table};
 #[doc(hidden)]
 /// `JoinDsl` support trait to emulate associated type constructors
 pub trait InternalJoinDsl<Rhs, Kind, On> {
-    type Output: AsQuery;
+    type Output;
 
     fn join(self, rhs: Rhs, kind: Kind, on: On) -> Self::Output;
 }
@@ -26,7 +26,7 @@ where
 /// `JoinDsl` support trait to emulate associated type constructors and grab
 /// the known on clause from the associations API
 pub trait JoinWithImplicitOnClause<Rhs, Kind> {
-    type Output: AsQuery;
+    type Output;
 
     fn join_with_implicit_on_clause(self, rhs: Rhs, kind: Kind) -> Self::Output;
 }

--- a/diesel/src/query_source/joins.rs
+++ b/diesel/src/query_source/joins.rs
@@ -5,6 +5,7 @@ use crate::expression::nullable::Nullable;
 use crate::expression::SelectableExpression;
 use crate::prelude::*;
 use crate::query_builder::*;
+use crate::query_dsl::InternalJoinDsl;
 use crate::result::QueryResult;
 use crate::sql_types::BoolOrNullableBool;
 use crate::util::TupleAppend;
@@ -272,6 +273,34 @@ where
         (rhs.source, rhs.on)
     }
 }
+
+impl<Lhs, Rhs, On> JoinTo<Rhs> for OnClauseWrapper<Lhs, On>
+where
+    Lhs: JoinTo<Rhs>,
+{
+    type FromClause = <Lhs as JoinTo<Rhs>>::FromClause;
+    type OnClause = <Lhs as JoinTo<Rhs>>::OnClause;
+
+    fn join_target(rhs: Rhs) -> (Self::FromClause, Self::OnClause) {
+        <Lhs as JoinTo<Rhs>>::join_target(rhs)
+    }
+}
+
+impl<Rhs, Kind, On1, On2, Lhs> InternalJoinDsl<Rhs, Kind, On1> for OnClauseWrapper<Lhs, On2>
+where
+    Lhs: InternalJoinDsl<Rhs, Kind, On1>,
+{
+    type Output = OnClauseWrapper<<Lhs as InternalJoinDsl<Rhs, Kind, On1>>::Output, On2>;
+
+    fn join(self, rhs: Rhs, kind: Kind, on: On1) -> Self::Output {
+        OnClauseWrapper {
+            source: self.source.join(rhs, kind, on),
+            on: self.on,
+        }
+    }
+}
+
+impl<Qs, On> QueryDsl for OnClauseWrapper<Qs, On> {}
 
 #[doc(hidden)]
 /// Convert any joins in a `FROM` clause into an inner join.

--- a/diesel_tests/tests/joins.rs
+++ b/diesel_tests/tests/joins.rs
@@ -467,6 +467,20 @@ fn selecting_parent_child_grandchild() {
         (sean.clone(), (posts[0].clone(), comments[2].clone())),
         (sean.clone(), (posts[2].clone(), comments[1].clone())),
     ];
+    assert_eq!(Ok(expected.clone()), data);
+
+    let data = users::table
+        .inner_join(
+            posts::table
+                .on(users::id.eq(posts::user_id).and(posts::id.eq(posts[0].id)))
+                .inner_join(comments::table),
+        )
+        .order((users::id, posts::id, comments::id))
+        .load(&connection);
+    let expected = vec![
+        (sean.clone(), (posts[0].clone(), comments[0].clone())),
+        (sean.clone(), (posts[0].clone(), comments[2].clone())),
+    ];
     assert_eq!(Ok(expected), data);
 
     let data = users::table

